src/Compilers/Test/Utilities/CSharp/NullableAttributesVisitor.cs

```diff
@@ -0,0 +1,213 @@
+        }
+
+        private void VisitList<TSymbol>(ImmutableArray<TSymbol> symbols) where TSymbol : Symbol
+        {
+            foreach (var symbol in symbols)
+            {
+                Visit(symbol);
+            }
+        }
+
+        private static string GetIndentString(Symbol symbol)
+        {
+            int level = 0;
+            while (true)
+            {
+                symbol = symbol.ContainingSymbol;
+                if (symbol.Kind == SymbolKind.Namespace)
+                {
+                    break;
+                }
+                level++;
+            }
+            return new string(' ', level * 4);
+        }
+
+        private static readonly SymbolDisplayFormat _displayFormat =
+            new SymbolDisplayFormat(
+                globalNamespaceStyle: SymbolDisplayGlobalNamespaceStyle.OmittedAsContaining,
+                typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces,
+                propertyStyle: SymbolDisplayPropertyStyle.ShowReadWriteDescriptor,
+                genericsOptions:
+                    SymbolDisplayGenericsOptions.IncludeTypeParameters |
+                    SymbolDisplayGenericsOptions.IncludeVariance |
+                    SymbolDisplayGenericsOptions.IncludeTypeConstraints,
+                memberOptions:
+                    SymbolDisplayMemberOptions.IncludeParameters |
+                    SymbolDisplayMemberOptions.IncludeType |
+                    SymbolDisplayMemberOptions.IncludeRef |
+                    SymbolDisplayMemberOptions.IncludeExplicitInterface,
+                parameterOptions:
+                    SymbolDisplayParameterOptions.IncludeOptionalBrackets |
+                    SymbolDisplayParameterOptions.IncludeDefaultValue |
+                    SymbolDisplayParameterOptions.IncludeParamsRefOut |
+                    SymbolDisplayParameterOptions.IncludeExtensionThis |
+                    SymbolDisplayParameterOptions.IncludeType |
+                    SymbolDisplayParameterOptions.IncludeName,
+                miscellaneousOptions:
+                    SymbolDisplayMiscellaneousOptions.EscapeKeywordIdentifiers |
+                    SymbolDisplayMiscellaneousOptions.UseErrorTypeSymbolName |
+                    SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier,
+                compilerInternalOptions:
+                    SymbolDisplayCompilerInternalOptions.UseMetadataMethodNames |
+                    SymbolDisplayCompilerInternalOptions.IncludeNonNullableTypeModifier);
+
+        private void ReportContainingSymbols(Symbol symbol)
+        {
+            if (symbol.Kind == SymbolKind.Namespace)
+            {
+                return;
+            }
+            if (_reported.Contains(symbol))
+            {
+                return;
+            }
+            ReportContainingSymbols(symbol.ContainingSymbol);
+            ReportSymbol(symbol, includeAlways: true);
```

Comment:
>, includeAlways: true [](start = 31, length = 21)

I didn't understand this part. We always want to report containing symbols?

Reply:
When we need to report a specific symbol, we first report all ancestors, from the outermost type down.

---
In reply to: [293589349](https://github.com/dotnet/roslyn/pull/36398#discussion_r293589349) [](ancestors = 293589349)